Prototype of passing server-provided fields to custom type converters

jskeet opened the pull request:
Would fix #5470, if adopted. Needs documentation if we go ahead.

Concerns:

- Is the Firestore team happy with this?
- Should we just work out a way of making the deserialization
  context public?
- Is it okay for the attribute to allow multiple and be inherited?
  (I can imagine a TypeConverterBase specifying it, for example...
  but what if we later have extra properties on it that conflict?)
- Is it okay for this to be very tightly bound to the implementation
  type? (There's no flexibility here - you need to know the keys
  at compile-time. I think that's okay, but...)

cc @schmidt-sebastian for Firestore, @ryandl for verification that it would solve their issue.

jskeet commented:
Have just noticed that we need to handle DeserializeValue as well as DeserializeMap... adding a second commit.

ryandle commented:
I have reviewed the changes and, Yes, this would definitely solve my scenario, thanks!

amanda-tarafa commented on apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreConverterConfigurationAttribute.cs:
```diff
@@ -0,0 +1,72 @@
+﻿// Copyright 2019, Google LLC
```
nit: 2020

jskeet replied:
Thanks, will fix.

amanda-tarafa reviewed (approved):
LGTM

jskeet commented:
At a Firestore team member suggestion, I've added an extra commit which moves the attribute to be a method-targeted attribute instead of a type-targeted attribute.

We'd want to rename the attribute and change the comments for it, but I wanted to present the idea first.

cc @ryandle 
cc @amanda-tarafa 
cc @schmidt-sebastian

I'm on the fence at the moment.

amanda-tarafa commented on apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreConverterConfigurationAttribute.cs:
```diff
@@ -0,0 +1,72 @@
+    /// Provides additional information for how Firestore converters (types implementing <see cref="IFirestoreConverter{T}"/>)
+    /// should be called.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
```
Why multiple?

jskeet replied:
No particular reason, and definitely something to ask ourselves. I think it makes sense if it's inherited, in that a subclass may want to specify it, then a superclass add it later... if it's specified in multiple inheritance levels, I think it makes sense to allow it to be specified multiple times at one level too. But I think we should decide "method or class" for sure, then make those decisions (and experiment with overrides etc).

amanda-tarafa reviewed (commented):
I think I like it being clearer that this is for deserialization only, althoguh that could also be achieved by changing the name of the class attribute to something like DesrializationConfig...
(Also, I seem to remember that overriding methods will inherit attributes, but not 100% certain about that though)

jskeet commented:
Okay, I've now rebased this to be a single commit that is method-targeted.

I've kept Multiple=true and specified Inherited=true - I think it makes sense to be able to compose the attribute however the user wants, basically.

I've added some basic documentation - I don't think we need a separate sample, although I could write one to match the existing one for document ID etc if we want. Let me know.

Happy to reconsider the attribute name, but otherwise this is good to go.

amanda-tarafa reviewed (approved):
LGTM. And agree with allowing full composition

jskeet commented:
Great - will merge this, then release a beta. If that works out for @ryandle we can then release a GA version in a week or so.
